src/Nethermind/Nethermind.AuRa/AuRaBlockFinalizationManager.cs

```diff
@@ -245,45 +306,74 @@ public void Dispose()
         private class ValidationStampCollection
         {
             private readonly IDictionary<Address, int> _validatorCount = new Dictionary<Address, int>();
-            [Todo("Optimization: circular sorted list?")]
-            private readonly SortedDictionary<long, Address> _blockValidator = new SortedDictionary<long, Address>();
-            
-            public int CountWith(BlockHeader block) => _validatorCount.ContainsKey(block.Beneficiary) ? _validatorCount.Count : _validatorCount.Count + 1;
+            private readonly Deque<BlockHeader> _blocks = new Deque<BlockHeader>();
+
+            public int Count => _validatorCount.Count;
 
             public void Add(BlockHeader blockHeader)
             {
-                if (!_blockValidator.ContainsKey(blockHeader.Number))
+                bool NotContainsBlock() => _blocks.Count == 0 || _blocks[0].Number > blockHeader.Number || _blocks[^1].Number < blockHeader.Number;
+
+                if (NotContainsBlock())
                 {
-                    _blockValidator[blockHeader.Number] = blockHeader.Beneficiary;
+                    if (_blocks.Count == 0 || _blocks[0].Number < blockHeader.Number)
+                    {
+                        _blocks.AddToFront(blockHeader);
+                    }
+                    else
+                    {
+                        _blocks.AddToBack(blockHeader);
+                    }
                     int count = _validatorCount.TryGetValue(blockHeader.Beneficiary, out count) ? count + 1 : 1;
                     _validatorCount[blockHeader.Beneficiary] = count;
                 }
             }
 
             public void RemoveAncestors(long blockNumber)
             {
```

Comment:
just curious - will it all be safe if there are JSON RPC replay block calls on a separate thread? and block producer?

Reply:
Yes, two things, this is just optimization and its compatible with non-consecutive calls, and secondly this is invoked based on BlockProcessed in BlockProcessor. Both BlockProducing and replaying blocks will have seperate blockprocessors than actual chain. They shouldn't even have hooked up finalization manger.